---
title: "ARVIN : Identifying Risk Noncoding Variants Using Disease-relevant Gene Regulatory Networks"
author: "Long Gao, Yasin Uzun, Kai Tan"
date: "October 31, 2017"
output: 
  pdf_document:
    toc: true
    toc_depth: 4
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1 Introcution

Identifying causalnoncoding variants remains a daunting task. Becausenoncoding variantsexert their effectsin the context of a gene regulatory network (GRN), we hypothesize that explicit use of disease-relevant GRNcan significantly improve the inference accuracy of noncoding risk variants. We describe Annotation of Regulatory Variants using Integrated Networks (ARVIN), a general computational framework for predictingcausal noncoding variants. For each disease, ARVIN first constructs a GRN using multi-dimensional omics data oncell/tissue-type relevant to the disease. ARVIN then uses a set of novel regulatory network-basedfeatures, combined with sequence-based features to make predictions.Using known causal variants ingene promoters and enhancersin a number of diseases, we show ARVIN outperforms state-of-the-art methods thatuse sequence-based features alone. 


## 2 Network construction
### 2.1 Enhancer prediction
ARVIN uses enhancer regions for mapping SNPs to enhancers. The data file must be in the following tab separated format:
#chromosome	enhancer_center	enhancer_probability  
chr6	138231000	0.86  
chr1	160807000	0.79  
chr1	160808600	0.88  
You can use CSI-ANN software for predicting the enhancers. You can download CSI-ANN from our lab web page http://tanlab4generegulation.org/CSIANNWebpage.html . CSI-ANN uses histone modification data as input to predict enhancer regions in genome-wide. You can use the output of CSI-ANN (which is as shown above) as input for ARVIN.
### 2.2 Enhancer-promoter interaction prediction
ARVIN uses enhancer-promoter interaction for mapping SNPS to genes via enhancers. The enhancer-promoter interaction data must be in the following format:
#Chr	Start	End	Target	Score  
chr9	22124001	22126001	ENST00000452276	0.93  
chr2	242792001	242794001	ENST00000485966	0.792  

### 2.3 Obtain gene-gene interaction network
### 2.3 Network scoring

## 3 Prepare features for risk variants prediction
### 3.1 Network-based features
#### 3.1.1 Betweenness centrality
#### 3.1.2 Closeness centrality
#### 3.1.3 Pagerank centrality
#### 3.1.4 Weighted degree
#### 3.1.5 Module score
### 3.2 GWAVA features
### 3.3 FunSeq features

## 4 Build a classifier for prioritizing risk varints
### 4.1 Train a random forest classifier
### 4.2 Predict causal disease variants
